### PR TITLE
Multi-factor AEM

### PR DESCRIPTION
**Pull request by phagara**

### CODE REVIEW

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal [hunk 1]
```diff
@@ -134,71 +131,180 @@ else
...
+        message "back to displaying the text secret (if provisioned)."
+        message "Otherwise, shut down the computer."
+        message ""
+        message "[ $(date) ] TOTP code: $(oathtool --totp -b "$seed")"
+        last=$(date +%s)
+
+        if plymouth_active; then
+            plymouth watch-keystroke --keys=$'\n' --command='touch /tmp/mfa-cont' &
+            plymouth watch-keystroke --keys=Qq --command='touch /tmp/mfa-skip' &
+            while [ ! -e /tmp/mfa-cont ] && [ ! -e /tmp/mfa-skip ]; do
```
Comment (rustybird) [inline anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: It's reasonable for this loop to run longer than systemd's `DefaultTimeoutStartSec` (90 seconds), so `TimeoutStartSec` should probably be bumped to something like `5m` in `-unseal.service`

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal [hunk 2]
```diff
@@ -134,71 +131,180 @@ else
...
+    if tpm_unsealdata $Z -i "$SEALED_SECRET" -o "$UNSEALED_SECRET" \
+            < "$SRK_PASSWORD_CACHE"; then
+        log "TOTP secret unsealed."
+        seed=$(cat "$UNSEALED_SECRET")
+
+        message "Please verify your OTP device displays the same TOTP code"
+        message "as shown below. Should the codes match, press <ENTER> to"
+        message "unseal your LUKS key file. If you do now have access to"
+        message "your OTP device but wish to continue despite the security"
+        message "implications, press <ENTER>. Pressing <Q> will revert"
```
Comment (rustybird) [inline anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: `<T>` seems more natural than `<Q>`

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal [hunk 3]
```diff
@@ -134,71 +131,180 @@ else
...
+if otp; then
+    log "Unsealing TOTP shared secret seed..."
+    if tpm_unsealdata $Z -i "$SEALED_SECRET" -o "$UNSEALED_SECRET" \
+            < "$SRK_PASSWORD_CACHE"; then
+        log "TOTP secret unsealed."
+        seed=$(cat "$UNSEALED_SECRET")
+
+        message "Please verify your OTP device displays the same TOTP code"
+        message "as shown below. Should the codes match, press <ENTER> to"
+        message "unseal your LUKS key file. If you do now have access to"
```
Comment (rustybird) [inline anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: `now` -> `not`

File: anti-evil-maid/README [hunk 4]
```diff
@@ -51,13 +51,19 @@ PCR-04: 93 33 4E 81 A6 9C 80 54 D6 87 C7 FD 76 7C 6F 4C 70 FC C6 73
...
 the article referenced above), certainly if you want to save the sealed secrets
 to your internal boot partition. In that case you SHOULD NOT pass the '-z'
 argument to tpm_takeownership.
 
+Please note that non-default owner password is required (ie. do not pass '-y'
+argument to tpm_takeownership)! Without an owner password set, the single-use
+anti-replay "freshness" tokens would be useless, as an attacker could simply
+rewrite the TPM NVRAM area with previously-valid-and-captured freshness token
+hash(es).
+
```
Comment (rustybird) [inline anti-evil-maid/README]: Now that `anti-evil-maid-install` handles taking ownership, let's get rid of this whole section 3 (except for the last paragraph).

File: anti-evil-maid/sbin/anti-evil-maid-lib [hunk 5]
```diff
@@ -54,6 +62,160 @@ waitfor() {
...
+checkfreshness() {
+    # check whether hash of an usealed freshness token (file path
+    # given as an argument) is contained in TPM NVRAM area
+    _hash=$(sha1sum "$1" | cut -d ' ' -f 1)
+    _lastslot=$((TPM_FRESHNESS_SLOTS - 1))
+    for _i in $(seq 0 $_lastslot); do
+        _slot=$(tpm_nvread_stdout -i "$TPM_FRESHNESS_INDEX" \
+            -n "$((_i * 20))" -s 20 | hex)
+        if [ "$_hash" == "$_slot" ]; then
+            message "Matching freshness token hash found in slot #$_i"
```
Comment (rustybird) [inline anti-evil-maid/sbin/anti-evil-maid-lib]: I think this message can be removed, it's a little noisy.

File: anti-evil-maid/sbin/anti-evil-maid-install [hunk 6]
```diff
@@ -30,6 +30,10 @@ Arguments:
 
   -F: passed on to mkfs.ext4 (don't ask for confirmation, etc.)
 
+  -m: set up a multi-factor auth AEM media (EXPERIMENTAL)
+      Using time-based one time password and a LUKS key file, provides resistance
```
Comment (rustybird) [inline anti-evil-maid/sbin/anti-evil-maid-install]: Please rewrap long lines like this one (and many of the `log` messages when taking into account the `anti-evil-maid-install: ` prefix) at max. 80 characters

File: anti-evil-maid/README [hunk 7]
```diff
@@ -191,5 +247,38 @@ and enter your disk decryption passphrase even though you can't see your
...
+media and forcing you to boot an unverified system).
+
+As a last resort, you can attempt to reset the TPM from BIOS (with similar effect
+to `resetfreshness`). Otherwise, it's game over.
+
+
+Someone saw my AEM text secret/LUKS passphrase
+----------------------------------------------
+
+You should've installed AEM in multi-factor mode. Better luck next time.
```
Comment (rustybird) [inline anti-evil-maid/README]: Users can at least add a new LUKS passphrase, revoke the old key slots, and reinstall AEM (with a different static text secret if needed). It's not a panacea for SSDs, but still.

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal (hunk 1 above)
Comment (phagara) [reply anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: Changed in 51dfc59.

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal (hunk 2 above)
Comment (phagara) [reply anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: Agreed, changed in 15d20e7.

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal (hunk 3 above)
Comment (phagara) [reply anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: Fixed in 0a2c2bf.

File: anti-evil-maid/README (hunk 4 above)
Comment (phagara) [reply anti-evil-maid/README]: Removed in e14ac93.

File: anti-evil-maid/sbin/anti-evil-maid-lib (hunk 5 above)
Comment (phagara) [reply anti-evil-maid/sbin/anti-evil-maid-lib]: Removed in 188f0db.

File: anti-evil-maid/README (hunk 7 above)
Comment (phagara) [reply anti-evil-maid/README]: I've broken that down into two scenarios and added recovery steps for each one separately. 9a76eb0

File: anti-evil-maid/sbin/anti-evil-maid-install (hunk 6 above)
Comment (phagara) [reply anti-evil-maid/sbin/anti-evil-maid-install]: Done for some other files, too. bedfa8f

Comment (phagara) [reply anti-evil-maid/sbin/anti-evil-maid-install]: Oh you mean including the logger prefix...

Comment (rustybird) [reply anti-evil-maid/sbin/anti-evil-maid-install]: There are more, see

    sed -e '/log "/!d' \
        -e 's/.*log "/anti-evil-maid-install: /' \
        -e 's/"$//' \
        anti-evil-maid-install \
    | grep -E '.{81,}'

Comment (phagara) [reply anti-evil-maid/sbin/anti-evil-maid-install]: Here it is: a051d2c. Sorry for misunderstanding (thought you meant line lengths in code, not script run logs).

Comment (rustybird) [reply anti-evil-maid/sbin/anti-evil-maid-install]: No problem. I'll test it some more tomorrow and the day after.

File: anti-evil-maid/README [hunk 8]
```diff
@@ -71,11 +60,19 @@ You should download the SINIT module required for your system.
...
-from:
+You can then download the module and unzip it. All the modules can be
+downloaded from:
 
 https://software.intel.com/protected-download/267276/183305
 
+Also, make sure you have the latest RACM update, if available (2nd & 3rd gen):
+https://software.intel.com/system/files/article/183305/intel-txt-sinit-acm-revocation-tools-guide-rev1-0_2.pdf
+
+It's possible to use 3rd gen SINIT/RACM on 2nd gen platforms. In fact, the
```
Comment (rustybird) [inline anti-evil-maid/README:84]: This is a really cool tip! It even made my T420 AEM compatible again - the 2nd gen blob doesn't pass the dreaded PCR sanity check anymore(?!), but the 3rd gen blob does...

Do you know if this applies to other Intel generations? Maybe all die shrinks?

Comment (phagara) [reply anti-evil-maid/README:84]: No clue, to be honest. Perhaps we should mention to use 3rd gen SINIT on 2nd gen CPUs if PCR sanity check fails (adding a "troubleshooting" readme section)?

**Comment by phagara** [comment]
Just realized it's possible to support even non-reseedable hardware TOTP tokens! User will simply need to store the manufacturer-provided seed (in a suitable format) in `/var/lib/anti-evil-maid/aem<.suffix>/secret.otp` and then (re-)run `anti-evil-maid-install`.

**Comment by rustybird** [comment]
> Just realized it's possible to support even non-reseedable hardware TOTP tokens!

Oh, nice!

File: anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal [hunk 9]
```diff
@@ -73,7 +73,7 @@ tcsd
 SEALED_SECRET_TXT=$TPM_DIR/$LABEL/secret.txt.sealed2
 SEALED_SECRET_KEY=$TPM_DIR/$LABEL/secret.key.sealed2
 SEALED_SECRET_OTP=$TPM_DIR/$LABEL/secret.otp.sealed2
-SEALED_SECRET_FSH=$TPM_DIR/$LABEL/secret.fsh.sealed2
+SEALED_SECRET_FSH=$TPM_DIR/$LABEL/secret.fre.sealed2
```
Comment (rustybird) [inline anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: It's also in the variable name

Comment (phagara) [reply anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal]: Thanks! Fixed by 72286d9.

**Comment by phagara** [comment]
Your GPG signatures disappeared from the first pull request against my repo as I rebase-merged it by mistake. Should I go back and fix that, then force-push the fully-signed history?

File: anti-evil-maid/sbin/anti-evil-maid-install [hunk 10]
```diff
@@ -229,6 +229,7 @@ if mfa && [ ! -e "$AEM_DIR/$LABEL/secret.otp" ]; then
         log "TOTP codes to be generated during AEM boot. Please fix"
         log "this by running (as root):"
         log "    timedatectl set-local-rtc 0"
+	waitforenter
```
Comment (rustybird) [inline anti-evil-maid/sbin/anti-evil-maid-install]: Stray tab. Just to clarify, it is not my intention to torture you.

**Comment by rustybird** [comment]
It's fine either way with me, I always diffed our trees and can vouch that everything's ok. But I don't know if Qubes policy requires all commits to be signed or only to last one (/cc @marmarek). If it's the former you'd have to re-sign all our commits with your key.

**Comment by rustybird** [comment]
(Although that would have the advantage that you can `--autosquash` a lot of commits)

File: anti-evil-maid/sbin/anti-evil-maid-install (hunk 10 above)
Comment (phagara) [reply anti-evil-maid/sbin/anti-evil-maid-install]: Heh, no problem at all. :) 6cd137a

**Comment by phagara** [comment]
Created a new branch with all commits (re-)signed, just in case. Also did the `--autosquash` thing where possible w/o manually resolving merge conflicts. Can you please check whether it's all OK? https://github.com/phagara/qubes-antievilmaid/commits/fix-sigs

**Comment by rustybird** [comment]
Sorry for the merge conflict! I messed up the order of the `-tpm-setup` fixups.

Commit ef0aedc looks a bit odd now, can you reword its message to:

```
-tpm-setup fixes

Must be run as root
(Re)start tcsd before tpm_takeownership call
Filter out tpm_takeownership's password prompts
```

If you start this second rebase a little deeper (on top of 5ad57a1), there are also 12 commits (revert vs. original) that can be filtered out.

**Comment by phagara** [comment]
Done & done. :)

**Comment by rustybird** [comment]
Looks like you force pushed a branch that still includes the revert/original commits.

**Comment by phagara** [comment]
Ah, sorry! Though `--autosquash` will get rid of those. :-\ Should be fixed now.

**Comment by rustybird** [comment]
Thanks! LGTM

**Comment by phagara** [comment]
What should be the next step? Open a second PR against your fork with the fixed signatures & squshed commits branch and close this one? Or should I file it directly against the upstream repo? (Plus, looks like I need to follow some GSoC [work product submission guidelines](https://developers.google.com/open-source/gsoc/help/work-product).)

**Comment by rustybird** [comment]
> What should be the next step? Open a second PR against your fork with the fixed signatures & squshed commits branch and  close this one? Or should I file it directly against the upstream repo?

Upstream repo. :))
You can close this PR.

> (Plus, looks like I need to follow some GSoC work product submission guidelines.)

Yes. If I understand it right - please double check! - it is sufficient to submit the URL of the new PR (which should have a detailed description and note the last commit pushed before the August 29 16:00 UTC deadline) to Google.

**Comment by phagara** [comment]
Alright, thanks!

**Pull request closed without merging**